File: 07_linked_list/07_10_sort_list/sort_list.js
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class SortList {
    sortList(head) {
        if (!head || !head.next) {
            return head;
        }

        // Find the middle of the list
        let slow = head;
        let fast = head.next;
        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Split the list into two halves
        let secondHalf = slow.next;
        slow.next = null;

        // Recursively sort the two halves
        let firstHalf = this.sortList(head);
        let secondHalfSorted = this.sortList(secondHalf);

        // Merge the sorted halves
        return this.merge(firstHalf, secondHalfSorted);
    }

    merge(first, second) {
        const union = new ListNode(0);
        let trav = union;

        while (first && second) {
            if (first.val < second.val) {
                trav.next = first;
                first = first.next;
            } else {
                trav.next = second;
                second = second.next;
            }
            trav = trav.next;
        }

        if (first) {
            trav.next = first;
        }
        if (second) {
            trav.next = second;
        }

        return union.next;
    }
}

function printList(label, head) {
    process.stdout.write(`${label} `);
    let current = head;
    while (current) {
        process.stdout.write(`${current.val} `);
        current = current.next;
    }
    console.log();
}

// Test Case 1: Random unsorted list
const head1 = new ListNode(4, new ListNode(2, new ListNode(1, new ListNode(3))));
printList("Test Case 1 (Before):", head1);
const sorter = new SortList();
const result1 = sorter.sortList(head1);
printList("Test Case 1 (After):", result1);  // Output: 1 -> 2 -> 3 -> 4

// Test Case 2: Already sorted list
const head2 = new ListNode(1, new ListNode(2, new ListNode(8, new ListNode(4))));
printList("Test Case 2 (Before):", head2);
const result2 = sorter.sortList(head2);
printList("Test Case 2 (After):", result2);  // Output: 1 -> 2 -> 4 -> 8

// Test Case 3: Single node
const head3 = new ListNode(5);
printList("Test Case 3 (Before):", head3);
const result3 = sorter.sortList(head3);
printList("Test Case 3 (After):", result3);  // Output: 5

File: 07_linked_list/07_10_sort_list/sort_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class SortList:
    def sortList(self, head):
        if not head or not head.next:
            return head

        # Find the middle of the list
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Split the list into two halves
        second_half = slow.next
        slow.next = None

        # Recursively sort the two halves
        first_half = self.sortList(head)
        second_half = self.sortList(second_half)

        # Merge the sorted halves
        return self.merge(first_half, second_half)

    def merge(self, first, second):
        union = ListNode(0)
        trav = union

        while first and second:
            if first.val < second.val:
                trav.next = first
                first = first.next
            else:
                trav.next = second
                second = second.next
            trav = trav.next

        if first:
            trav.next = first
        if second:
            trav.next = second

        return union.next

def print_list(label, head):
    print(f"{label} ", end="")
    current = head
    while current:
        print(current.val, end=" ")
        current = current.next
    print()

# Test Case 1: Random unsorted list
head1 = ListNode(4, ListNode(2, ListNode(1, ListNode(3))))
print_list("Test Case 1 (Before):", head1)
sorter = SortList()
result1 = sorter.sortList(head1)
print_list("Test Case 1 (After):", result1)  # Output: 1 -> 2 -> 3 -> 4

# Test Case 2: Already sorted list
head2 = ListNode(1, ListNode(2, ListNode(8, ListNode(4))))
print_list("Test Case 2 (Before):", head2)
result2 = sorter.sortList(head2)
print_list("Test Case 2 (After):", result2)  # Output: 1 -> 2 -> 4 -> 8

# Test Case 3: Single node
head3 = ListNode(5)
print_list("Test Case 3 (Before):", head3)
result3 = sorter.sortList(head3)
print_list("Test Case 3 (After):", result3)  # Output: 5
